### PR TITLE
Add missing compression argument in to_parquet underlying methods

### LINKED ISSUE #2907: dask dataframe to_parquet method with engine=pyarrow fails trying to write metadata
(version 0.16)

dask dataframe to_parquet method seems to fail because at some point the compression argument becomes part of kwargs.

    def _write_partition_pyarrow(df, open_with, filename, write_index, compression, metadata_path=None, **kwargs):
        import pyarrow as pa
        from pyarrow import parquet
        t = pa.Table.from_pandas(df, preserve_index=write_index)

        with open_with(filename, 'wb') as fil:
            parquet.write_table(t, fil, **kwargs)

        if metadata_path is not None:
            with open_with(metadata_path, 'wb') as fil:
                parquet.write_metadata(t.schema, fil, **kwargs)

The method above belongs to io/parquet.py (from dask project module), and as you can see, **kwargs is used in parquet.write_table(t, fil, **kwargs), and  parquet.write_metadata(t.schema, fil, **kwargs). However, parquet.write_metadata does not expect a compression argument. In fact the method signature is the following:

    def write_metadata(schema, where, version='1.0', use_deprecated_int96_timestamps=False,
                   coerce_timestamps=None)

the signature comes from pyarrow/parquet.py module (from pyarrow project)

### PR DESCRIPTION
**Pull request by castalheiro**
Fixes #2907.

- [x] Tests added / passed
- [x] Passes `flake8 dask`
- [ ] Fully documented, including `docs/source/changelog.rst` for all changes
      and one of the `docs/source/*-api.rst` files for new API

### CODE REVIEW

File: dask/dataframe/io/parquet.py [hunk 1]
```diff
@@ -397,22 +394,23 @@ def _write_pyarrow(df, path, write_index=None, append=False,
     template = fs.sep.join([path, 'part.%i.parquet'])
 
     write = delayed(_write_partition_pyarrow)
-    first_kwargs = kwargs.copy()
-    first_kwargs['metadata_path'] = fs.sep.join([path, '_metadata'])
-    writes = [write(part, open_with, template % i, write_index,
-                    **(kwargs if i else first_kwargs))
+    metadata_path = fs.sep.join([path, '_metadata'])
+    writes = [write(part, open_with, template % i, write_index, compression, metadata_path,
+                    **kwargs)
```
Comment (jcrist) [inline dask/dataframe/io/parquet.py]: Please revert the handling of `first_kwargs`/`metadata_path` here - the way you changed it the metadata is written once for every task instead of just once by the first task.

File: dask/dataframe/io/parquet.py [hunk 2]
```diff
@@ -397,22 +394,23 @@ def _write_pyarrow(df, path, write_index=None, append=False,
...
                              metadata_path=None, **kwargs):
     import pyarrow as pa
     from pyarrow import parquet
     t = pa.Table.from_pandas(df, preserve_index=write_index)
 
     with open_with(filename, 'wb') as fil:
-        parquet.write_table(t, fil, **kwargs)
+        kwargs_copy = kwargs.copy()
+        kwargs_copy['compression'] = compression
+        parquet.write_table(t, fil, **kwargs_copy)
```
Comment (jcrist) [inline dask/dataframe/io/parquet.py]: You can just do `parquet.write_table(t, fil, compression=compression, **kwargs)`

File: dask/dataframe/io/parquet.py [hunk 3]
```diff
@@ -540,7 +538,7 @@ def read_parquet(path, columns=None, filters=None, categories=None, index=None,
                 categories=categories, index=index)
 
 
-def to_parquet(df, path, engine='auto', compression='default', write_index=None,
+def to_parquet(df, path, engine='auto', compression=None, write_index=None,
```
Comment (jcrist) [inline dask/dataframe/io/parquet.py]: Please revert this change, `compression='default'` means the default compression for the selected backend (see the docstring below). `None` means something else.

File: dask/dataframe/io/parquet.py [hunk 4]
```diff
@@ -397,22 +394,23 @@ def _write_pyarrow(df, path, write_index=None, append=False,
...
-                    **(kwargs if i else first_kwargs))
+    metadata_path = fs.sep.join([path, '_metadata'])
+    writes = [write(part, open_with, template % i, write_index, compression, metadata_path,
+                    **kwargs)
               for i, part in enumerate(df.to_delayed())]
     return delayed(writes)
 
 
-def _write_partition_pyarrow(df, open_with, filename, write_index,
+def _write_partition_pyarrow(df, open_with, filename, write_index, compression=None,
```
Comment (jcrist) [inline dask/dataframe/io/parquet.py]: I'd handle this fix by reverting this change (leave `compression` in `kwargs`, and do:

`kwargs.pop('compression')` before passing to `write_metadata`. This ensures that compression only gets passed to the backend if it's specified (so that `'default'` still works, and removes it from being passed to `write_metadata`.

File: dask/dataframe/io/parquet.py (hunk 1 above)
Comment (castalheiro) [reply dask/dataframe/io/parquet.py]: done

File: dask/dataframe/io/parquet.py (hunk 4 above)
Comment (castalheiro) [reply dask/dataframe/io/parquet.py]: done

File: dask/dataframe/io/parquet.py (hunk 3 above)
Comment (castalheiro) [reply dask/dataframe/io/parquet.py]: done

File: dask/dataframe/io/parquet.py (hunk 4 above)
Comment (castalheiro) [reply dask/dataframe/io/parquet.py]: done

**Comment by castalheiro** [comment]
i'm trying to write the tests, but how do i check a parquet file compression after reading it?

**Comment by jcrist** [comment]
That's hard to check without explicitly looking at the thrift metadata explicitly (which requires fastparquet, and still isn't the easiest to do). Instead I'd just test that you can roundtrip write/read with the compression specified.

File: dask/dataframe/io/parquet.py [hunk 5]
```diff
@@ -416,6 +413,8 @@ def _write_partition_pyarrow(df, open_with, filename, write_index,
 
     if metadata_path is not None:
         with open_with(metadata_path, 'wb') as fil:
+            if 'compression' in kwargs:
+                kwargs.pop('compression')
```
Comment (jcrist) [inline dask/dataframe/io/parquet.py]: Not a big deal, but this might be slightly cleaner as either:

```python
if 'compression' in kwargs:
    del kwargs['compression']
```

or 

```python
kwargs.pop('compression', None)
```

I'd probably do the second, but I'll leave that up to you.

File: dask/tests/test_parquet.py [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+
+    return parquet_file_path
+'''
+
+
+@pytest.mark.parametrize('compression, engine', [
+    (None, 'pyarrow'), (None, 'fastparquet'),
+    ('SNAPPY', 'pyarrow'), ('SNAPPY', 'fastparquet'),
+    ('GZIP', 'pyarrow'), ('GZIP', 'fastparquet')])
+def test_writing_parquet_with_gzip_compression(compression, engine):
```
Comment (jcrist) [inline dask/tests/test_parquet.py]: We already have a test file for this in `dask/dataframe/io/tests/test_parquet.py`. Please move this test there.

File: setup.py [hunk 7]
```diff
@@ -25,7 +25,7 @@
 
 # Only include pytest-runner in setup_requires if we're invoking tests
 if {'pytest', 'test', 'ptr'}.intersection(sys.argv):
-    setup_requires = ['pytest-runner']
+    setup_requires = ['pytest-runner', 'fastparquet', 'pyarrow']
```
Comment (jcrist) [inline setup.py]: Please revert this. `setup_requires` is for installing dependencies that are required for building the package (of which neither of these are).

File: dask/tests/test_parquet.py (hunk 6 above)
Comment (jcrist) [reply dask/tests/test_parquet.py]: Also note that in that file there is a fixture for `engine` that skips tests requiring that engine if it's not installed. There's also a fixture for handling temporary directories. Your test would then look something like (not tested):

```python
@pytest.mark.parametrize('compression', ['default', None, 'snappy', 'gzip'])
def test_to_parquet_compression(tmpdir, engine, compression):
    fn = str(tmpdir)
    if engine == 'fastparquet' and compression == 'snappy':
        pytest.importorskip('snappy')
    df = pd.DataFrame({'x': ['a', 'b', 'c'] * 10, 'y': [1, 2, 3] * 10})
    ddf = dd.from_pandas(df, npartitions=3)
    ddf.to_parquet(fn, engine=engine, compression=compression)
    ddf2 = dd.read_parquet(fn, engine=engine)
    assert_eq(ddf2, df)
```

Also note that `snappy` isn't a required dependency of fastparquet, and may not be installed. `pyarrow` bundles `libsnappy`, so there's no fear of it missing there. I'd either check if snappy is installed and skip if not (as done above), or just not test with `snappy`. Up to you.

File: setup.py (hunk 7 above)
Comment (castalheiro) [reply setup.py]: done

File: dask/dataframe/io/tests/test_parquet.py [hunk 8]
```diff
@@ -597,3 +596,21 @@ def test_parquet_select_cats(tmpdir):
...
+def test_writing_parquet_with_compression(tmpdir, compression):
+    fn = str(tmpdir)
+
+    df = pd.DataFrame({'x': ['a', 'b', 'c'] * 10, 'y': [1, 2, 3] * 10})
+    ddf = dd.from_pandas(df, npartitions=3)
+
+    check_fastparquet()
+    ddf.to_parquet(fn, compression=compression, engine='fastparquet')
+    out = dd.read_parquet(fn, engine='fastparquet')
+    assert_eq(out, ddf)
```
Comment (jcrist) [inline dask/dataframe/io/tests/test_parquet.py]: It would be better if you used the `engine` fixture as I posted before. As is, both the fastparquet and `arrow` test uses the same temporary directory, and the arrow tests will only run if fastparquet is installed (since that's checked first).

```python
def test_writing_parquet_with_compression(tmpdir, compression, engine):
   ...
```

File: dask/dataframe/io/tests/test_parquet.py [hunk 9]
```diff
@@ -597,3 +596,18 @@ def test_parquet_select_cats(tmpdir):
...
+
+    if engine == 'fastparquet' and compression == 'snappy':
+        pytest.importorskip('snappy')
+
+    df = pd.DataFrame({'x': ['a', 'b', 'c'] * 10, 'y': [1, 2, 3] * 10})
+    ddf = dd.from_pandas(df, npartitions=3)
+
+    ddf.to_parquet(fn, compression=compression, engine=engine)
+    out = dd.read_parquet(fn, engine=engine)
+    assert_eq(out, ddf)
```
Comment (jcrist) [inline dask/dataframe/io/tests/test_parquet.py]: `assert_eq` assumes one of the arguments is a pandas dataframe. You probably meant `df` instead of `ddf`.

Comment (jcrist) [reply dask/dataframe/io/tests/test_parquet.py]: Note that you can run the tests locally instead of waiting for travis.

```
# assumes all dependencies are already installed
py.test dask/dataframe/io/tests/test_parquet.py -k test_writing_parquet_with_compression
```

Comment (castalheiro) [reply dask/dataframe/io/tests/test_parquet.py]: All parquet tests fail on my machine. I'll change the data frame input parameter and hope it goes ok

Comment (jcrist) [reply dask/dataframe/io/tests/test_parquet.py]: > All parquet tests fail on my machine.

Please check that you have the proper versions of pyarrow/fastparquet, and their dependencies installed.

**Comment by jcrist** [comment]
@castalheiro, I fixed the failing test, should be good to go once tests pass on travis/appveyor. Thanks for the fix.

---

For future reference, it's generally not advised to submit PRs from your master branch (as you've done here), but rather to make a branch and push your fixes there. This allows you to keep working, as your fix doesn't occupy your master branch. In general most people follow a workflow similar to the one described here: http://www.asmeurer.com/git-workflow/.
